filters/diag/logbody.go

```diff
@@ -0,0 +1,139 @@
+package diag
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/filters/flowid"
+)
+
+type logBody struct {
+	limit    int
+	request  bool
+	response bool
+}
+
+// NewLogBody creates a filter specification for the 'logBody()' filter.
+func NewLogBody() filters.Spec { return logBody{} }
+
+// Name returns the logBody filtern name.
+func (logBody) Name() string {
+	return filters.LogBodyName
+}
+
+func (logBody) CreateFilter(args []interface{}) (filters.Filter, error) {
+	var (
+		request  = false
+		response = false
+	)
+
+	if len(args) != 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	opt, ok := args[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	switch opt {
+	case "response":
+		response = true
+	case "request":
+		request = true
+	default:
+		return nil, fmt.Errorf("failed to match %q: %w", opt, filters.ErrInvalidFilterParameters)
+	}
+
+	limit, ok := args[1].(float64)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert to float64: %w", filters.ErrInvalidFilterParameters)
+	}
+
+	return &logBody{
+		limit:    int(limit),
+		request:  request,
+		response: response,
+	}, nil
+}
+
+func (lb *logBody) Request(ctx filters.FilterContext) {
+	if !lb.request {
+		return
+	}
+
+	req := ctx.Request()
+	if req.Body != nil {
+		req.Body = LogBodyStream(
+			lb.limit,
+			func(chunk []byte) {
+				ctx.Logger().Infof(
+					`logBody("request") %s: %s`,
+					req.Header.Get(flowid.HeaderName),
+					chunk)
+			},
+			req.Body,
+		)
+	}
+}
+
+func (lb *logBody) Response(ctx filters.FilterContext) {
+	if !lb.response {
+		return
+	}
+
+	rsp := ctx.Response()
+	if rsp.Body != nil {
+		rsp.Body = LogBodyStream(
+			lb.limit,
+			func(chunk []byte) {
+				ctx.Logger().Infof(
+					`logBody("response") %s: %s`,
+					ctx.Request().Header.Get(flowid.HeaderName),
+					chunk)
+			},
+			rsp.Body,
+		)
+	}
+}
+
+type logBodyStream struct {
+	limit int
+	f     func([]byte)
+	input io.ReadCloser
+}
+
+func newLogBodyStream(limit int, f func([]byte), rc io.ReadCloser) io.ReadCloser {
+	return &logBodyStream{
+		limit: limit,
+		f:     f,
+		input: rc,
+	}
+}
+
+func (lb *logBodyStream) Read(p []byte) (n int, err error) {
+	if lb.limit <= 0 {
+		return lb.input.Read(p)
+	}
+
+	n, err = lb.input.Read(p)
+	//println("n:", n, ", err: ", err.Error())
+	if n > 0 {
+		if len(p) > lb.limit {
+			lb.f(p[0:lb.limit])
+		} else {
+			lb.f(p)
+		}
```

Comment:
Maybe
```go
lb.f(p[:min(len(p), lb.limit)])
```

Reply:
requires go 1.21, but I guess it's fine to change go.mod to require that :)